### PR TITLE
feat(auth): deleting a user from the ui

### PR DESCRIPTION
**Pull request by rBergCode**
# Summary
The objective was to implement the delete account feature to let user delete their account.

# Details
I also did some useful changes:

- Move the collection name of the profile repository to a companion object.
- Implemented a `isEmailProvider` to know wether the user is connected via email or google
- Kind of fixed reliability of the `testWholeNavigation` test

### CODE REVIEW

File: app/src/main/java/ch/hikemate/app/model/authentication/AuthRepository.kt [hunk 1]
```diff
@@ -71,4 +72,29 @@ interface AuthRepository {
...
+      onSuccess: () -> Unit = {},
+      onErrorAction: (Exception) -> Unit
+  )
+
+  /**
+   * Return true if the user is signed in with an email provider.
+   *
+   * @param user The FirebaseUser object.
+   */
+  fun isEmailProvider(user: FirebaseUser): Boolean
```
Comment (Wylwi) [inline app/src/main/java/ch/hikemate/app/model/authentication/AuthRepository.kt:99]: Nitpick: Sorry about this, really nitpicking. Could you elaborate about what is an "email provider"? Strictly speaking, Facebook Auth provides email, so does Google, and so on... It might be clear when you're doing this, but for a newcomer, it's not easily understandable

File: app/src/main/java/ch/hikemate/app/model/authentication/AuthViewModel.kt [hunk 2]
```diff
@@ -148,4 +149,36 @@ class AuthViewModel(
...
+        activity = activity,
+        onSuccess = {
+          _currentUser.value = null
+          onSuccess()
+        },
+        onErrorAction = onErrorAction)
+  }
+
+  /** Returns whether the current user is connect with an email provider. */
+  fun isEmailProvider(): Boolean {
```
Comment (Wylwi) [inline app/src/main/java/ch/hikemate/app/model/authentication/AuthViewModel.kt:177]: Ditto: Same as above about email provider

File: app/src/main/java/ch/hikemate/app/model/authentication/FirebaseAuthRepository.kt [hunk 3]
```diff
@@ -127,4 +134,91 @@ class FirebaseAuthRepository : AuthRepository {
...
+
+            // Delete the user profile and saved hikes
+            FirebaseFirestore.getInstance()
+                .collection(ProfileRepositoryFirestore.PROFILES_COLLECTION)
+                .document(user.uid)
+                .delete()
+            FirebaseFirestore.getInstance()
+                .collection(SavedHikesRepositoryFirestore.SAVED_HIKES_COLLECTION)
+                .document(user.uid)
+                .delete()
```
Comment (Wylwi) [inline app/src/main/java/ch/hikemate/app/model/authentication/FirebaseAuthRepository.kt:165]: Question: Do you have an easy way to check for correctness of the data deletion? I'm pretty sure Firebase should return a boolean somewhere to indicate that deletion has correctly happened

Comment (Wylwi) [reply app/src/main/java/ch/hikemate/app/model/authentication/FirebaseAuthRepository.kt:165]: Part of my concern is about the deletion of the Firebase user (and thus permission) before the deletion is finished, leading to permissions errors

File: app/src/main/java/ch/hikemate/app/model/authentication/FirebaseAuthRepository.kt [hunk 4]
```diff
@@ -127,4 +134,91 @@ class FirebaseAuthRepository : AuthRepository {
...
+          },
+          { onErrorAction(Exception("Error re-authenticating user")) })
+    } else {
+      onErrorAction(Exception("No user is currently signed in"))
+    }
+  }
+
+  override fun isEmailProvider(user: FirebaseUser): Boolean {
+    val providers = user.providerData.map { it.providerId }
+    return providers.contains(EmailAuthProvider.PROVIDER_ID)
```
Comment (Wylwi) [inline app/src/main/java/ch/hikemate/app/model/authentication/FirebaseAuthRepository.kt:187]: Nitpicking: Don't you have a more performance efficient way to do it? Currently, this implementation is θ(n^2) in CPU usage (not accounting for comparison). I'm confident you can check for this in a θ(n) way

File: app/src/main/java/ch/hikemate/app/model/authentication/FirebaseAuthRepository.kt [hunk 5]
```diff
@@ -127,4 +134,91 @@ class FirebaseAuthRepository : AuthRepository {
...
+          .addOnFailureListener { e ->
+            Log.e("DeleteAccount", "Re-authentication failed: ${e.message}")
+            onFailure()
+          }
+          .addOnSuccessListener { result ->
+            Log.d("DeleteAccount", "Re-authentication successful: $result")
+            onSuccess()
+          }
+    }
+  }
```
Comment (Wylwi) [inline app/src/main/java/ch/hikemate/app/model/authentication/FirebaseAuthRepository.kt:223]: Praise: Nice and efficient way to do a reauthentication!

File: app/src/main/java/ch/hikemate/app/model/profile/ProfileRepositoryFirestore.kt
Comment (Wylwi) [inline app/src/main/java/ch/hikemate/app/model/profile/ProfileRepositoryFirestore.kt:1]: Praise: Nice refactor!

File: app/src/androidTest/java/ch/hikemate/app/ui/profile/DeleteAccountScreenTest.kt [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  private lateinit var navigationActions: NavigationActions
+
+  @get:Rule val composeTestRule = createComposeRule()
+
+  @Before
+  fun setUp() {
+    navigationActions = mockk(relaxed = true)
+    authViewModel = mockk(relaxed = true)
+
+    every { navigationActions.goBack() } returns Unit
```
Comment (Wylwi) [inline app/src/androidTest/java/ch/hikemate/app/ui/profile/DeleteAccountScreenTest.kt:30]: Question: Why is this needed? Is it an issue if we go back?

File: app/src/test/java/ch/hikemate/app/authentication/FirebaseAuthRepositoryTest.kt [hunk 7]
```diff
@@ -304,4 +336,48 @@ class FirebaseAuthRepositoryTest {
...
+      every { mockVoidTask.isSuccessful } returns true
+      every { mockFirebaseAuth.currentUser } returns mockFirebaseUser
+      every { mockFirebaseUser.providerData } returns
+          listOf(mockk { every { providerId } returns EmailAuthProvider.PROVIDER_ID })
+      every { mockDocument.delete() } returns mockVoidTask
+
+      val onSuccess: () -> Unit = mockk(relaxed = true)
+      val mockActivity: Activity = mockk(relaxed = true)
+
+      repository.deleteAccount("password", mockActivity, onSuccess, mockOnError)
```
Comment (Wylwi) [inline app/src/test/java/ch/hikemate/app/authentication/FirebaseAuthRepositoryTest.kt:352]: Question: Do we agree that in this setup (i.e. with those mocks), "password" will never be checked for correctness?

File: app/src/test/java/ch/hikemate/app/authentication/FirebaseAuthRepositoryTest.kt
Comment (Wylwi) [inline app/src/test/java/ch/hikemate/app/authentication/FirebaseAuthRepositoryTest.kt:1]: Praise: You did an amazing job on testing this repo!!

Comment (Wylwi) [inline app/src/test/java/ch/hikemate/app/authentication/FirebaseAuthRepositoryTest.kt:1]: Nitpick: Could you add one or two tests where the tests check for failure please?

File: app/src/test/java/ch/hikemate/app/authentication/AuthViewModelTest.kt [hunk 8]
```diff
@@ -361,4 +362,50 @@ class AuthViewModelTest {
...
+    doAnswer { arguments ->
+          val onSuccess = arguments.getArgument<() -> Unit>(2)
+          onSuccess()
+          null
+        }
+        .`when`(mockRepository)
+        .deleteAccount(any(), any(), any(), any())
+
+    // Verify that currentUser is initially mockFirebaseUser
+    assertEquals(mockFirebaseUser, viewModel.currentUser.first())
```
Comment (Wylwi) [inline app/src/test/java/ch/hikemate/app/authentication/AuthViewModelTest.kt:382]: Question: Why do you use `.first()` here and not `value`? `.first()`, even if it might be working, doesn't look appropriated for a single user

File: app/src/test/java/ch/hikemate/app/authentication/AuthViewModelTest.kt [hunk 9]
```diff
@@ -361,4 +362,50 @@ class AuthViewModelTest {
...
+    setupSignedInUser()
+
+    // Simulate a successful call to isEmailProvider
+    `when`(mockRepository.isEmailProvider(any())).thenReturn(true)
+
+    viewModel.isEmailProvider()
+
+    // Verify that the repository's isEmailProvider was called
+    verify(mockRepository).isEmailProvider(any())
+  }
```
Comment (Wylwi) [inline app/src/test/java/ch/hikemate/app/authentication/AuthViewModelTest.kt:410]: Praise: Simple but efficient test!

File: app/src/test/java/ch/hikemate/app/authentication/AuthViewModelTest.kt
Comment (Wylwi) [inline app/src/test/java/ch/hikemate/app/authentication/AuthViewModelTest.kt:1]: Nitpick: Could you add one or two tests where the tests check for failure please?

File: app/src/androidTest/java/ch/hikemate/app/ui/profile/DeleteAccountScreenTest.kt [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+  @Test
+  fun isEverythingDisplayed() {
+    setUpActivityEmailProvider()
+    composeTestRule.onNodeWithTag(BackButton.BACK_BUTTON_TEST_TAG).assertIsDisplayed()
+    composeTestRule.onNodeWithTag(DeleteAccountScreen.TEST_TAG_TITLE).assertIsDisplayed()
+    composeTestRule.onNodeWithTag(DeleteAccountScreen.TEST_TAG_INFO_TEXT).assertIsDisplayed()
+    composeTestRule
+        .onNodeWithTag(DeleteAccountScreen.TEST_TAG_DELETE_ACCOUNT_BUTTON)
+        .assertIsDisplayed()
+  }
```
Comment (Wylwi) [inline app/src/androidTest/java/ch/hikemate/app/ui/profile/DeleteAccountScreenTest.kt:52]: Warning: This doesn't check for `TEST_TAG_PASSWORD_INPUT`

File: app/src/main/java/ch/hikemate/app/ui/profile/DeleteAccountScreen.kt [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+ *
+ * @param navigationActions The navigation actions.
+ * @param authViewModel The authentication view model.
+ */
+@Composable
+fun DeleteAccountScreen(navigationActions: NavigationActions, authViewModel: AuthViewModel) {
+  val context = LocalContext.current
+
+  val passwordMustNotBeEmptyError =
+      stringResource(R.string.delete_account_password_must_be_filled_error)
```
Comment (Wylwi) [inline app/src/main/java/ch/hikemate/app/ui/profile/DeleteAccountScreen.kt:55]: Question: Why do you have this string here and all the other ones in the composable tree? Could you maybe normalize it, in the way you want?

File: app/src/main/java/ch/hikemate/app/ui/profile/DeleteAccountScreen.kt [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+                  TextSelectionColors(
+                      handleColor = primaryColor,
+                      backgroundColor = primaryColor,
+                  ))
+
+  var password by remember { mutableStateOf("") }
+
+  Column(
+      modifier =
+          Modifier.testTag(Screen.SIGN_IN_WITH_EMAIL)
```
Comment (Wylwi) [inline app/src/main/java/ch/hikemate/app/ui/profile/DeleteAccountScreen.kt:74]: Error: You put the wrong test tag here

File: app/src/main/java/ch/hikemate/app/ui/profile/DeleteAccountScreen.kt [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+
+        if (authViewModel.isEmailProvider())
+            OutlinedTextField(
+                modifier =
+                    Modifier.fillMaxWidth().testTag(DeleteAccountScreen.TEST_TAG_PASSWORD_INPUT),
+                visualTransformation = PasswordVisualTransformation(),
+                colors = inputColors,
+                value = password,
+                onValueChange = { password = it },
+                label = { Text(stringResource(R.string.delete_account_password_label)) })
```
Comment (Wylwi) [inline app/src/main/java/ch/hikemate/app/ui/profile/DeleteAccountScreen.kt:101]: Praise: Clean way to do it!

File: app/src/androidTest/java/ch/hikemate/app/ui/profile/DeleteAccountScreenTest.kt
Comment (Wylwi) [inline app/src/androidTest/java/ch/hikemate/app/ui/profile/DeleteAccountScreenTest.kt:1]: Question: Why is this file in the profile folder and not in the auth folder? Is it arbitrary or is there any reason?

File: app/src/androidTest/java/ch/hikemate/app/ui/profile/DeleteAccountScreenTest.kt [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+  @Test
+  fun checkCanDeleteAccount() {
+    setUpActivityEmailProvider()
+
+    composeTestRule
+        .onNodeWithTag(DeleteAccountScreen.TEST_TAG_PASSWORD_INPUT)
+        .performTextInput("password")
+    composeTestRule.onNodeWithTag(DeleteAccountScreen.TEST_TAG_DELETE_ACCOUNT_BUTTON).performClick()
+
+    verify(exactly = 1) { authViewModel.deleteAccount(any(), any(), any(), any()) }
```
Comment (Wylwi) [inline app/src/androidTest/java/ch/hikemate/app/ui/profile/DeleteAccountScreenTest.kt:83]: Suggestion: This is not required, but I think it would be a great improvement to check that the authViewModel get called with the correct password.
Below is a suggestion done without testing. Syntax might not be correct and the "password" should be put in a constant
```suggestion
    verify(exactly = 1) { authViewModel.deleteAccount(eq("password"), any(), any(), any()) }
    verify(exactly = 0) { authViewModel.deleteAccount(not(eq("password")), any(), any(), any()) }
```

File: app/src/androidTest/java/ch/hikemate/app/ui/profile/DeleteAccountScreenTest.kt [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  @Test
+  fun checkCanDeleteAccountWithoutPassword() {
```
Comment (Wylwi) [inline app/src/androidTest/java/ch/hikemate/app/ui/profile/DeleteAccountScreenTest.kt:87]: Nitpick: The name of the test is (in my understanding) not corresponding to the actual behavior of the test

**Review by Wylwi — CHANGES_REQUESTED** [review]
I'm not going to lie, you did an awesome job on this. I am just asking for a few changes to be done, some nitpicking and a lot of questions! Your work is impressive overall!

File: app/src/androidTest/java/ch/hikemate/app/navigation/HikeMateAppNavigationTest.kt [hunk 16]
```diff
@@ -38,8 +38,8 @@ class HikeMateAppNavigationTest {
 
   @Before
   fun setUp() {
-    composeTestRule.setContent { HikeMateApp() }
     auth.signOut()
+    composeTestRule.setContent { HikeMateApp() }
```
[the remark this answers is not in the document]
Comment (rBergCode) [reply app/src/androidTest/java/ch/hikemate/app/navigation/HikeMateAppNavigationTest.kt:42]: I did this as I though that it would make the tests pass more accurately. Actually it doesn't but in #159 you will remove this test anyway so I will merge it as soon as you merge it to main!

File: app/src/main/java/ch/hikemate/app/model/authentication/FirebaseAuthRepository.kt (hunk 4 above)
Comment (rBergCode) [reply app/src/main/java/ch/hikemate/app/model/authentication/FirebaseAuthRepository.kt:187]: Actually it's θ(n) not θ(n^2) but I will change it for θ(n)

File: app/src/androidTest/java/ch/hikemate/app/ui/profile/DeleteAccountScreenTest.kt (hunk 10 above)
Comment (rBergCode) [reply app/src/androidTest/java/ch/hikemate/app/ui/profile/DeleteAccountScreenTest.kt:52]: That's true! It's because the password field would not appear if the user is signed in with Google. That's why I added 2 more tests to test this only. Do you think I should still add it here ?

File: app/src/androidTest/java/ch/hikemate/app/ui/profile/DeleteAccountScreenTest.kt
Comment (rBergCode) [reply app/src/androidTest/java/ch/hikemate/app/ui/profile/DeleteAccountScreenTest.kt:1]: It is as the delete process is part of the profile page (from the `ProfileScreen` -> `DeleteAccountScreen`)

File: app/src/androidTest/java/ch/hikemate/app/ui/profile/DeleteAccountScreenTest.kt (hunk 6 above)
Comment (rBergCode) [reply app/src/androidTest/java/ch/hikemate/app/ui/profile/DeleteAccountScreenTest.kt:30]: The issue is because as we set the content of the test activity to be our `DeleteAccountScreen`, there are no previous page. Going back would then throw an error

File: app/src/test/java/ch/hikemate/app/authentication/FirebaseAuthRepositoryTest.kt (hunk 7 above)
Comment (rBergCode) [reply app/src/test/java/ch/hikemate/app/authentication/FirebaseAuthRepositoryTest.kt:352]: What do you mean by "will never be checked for correctness? It won't do any request to Firebase for sure. So If i understand your question correctly I would say yes.

File: app/src/test/java/ch/hikemate/app/authentication/AuthViewModelTest.kt (hunk 8 above)
Comment (rBergCode) [reply app/src/test/java/ch/hikemate/app/authentication/AuthViewModelTest.kt:382]: Actually, I took this from this file's others tests. I will change them all to `.value`. Nice catch!

File: app/src/main/java/ch/hikemate/app/ui/profile/DeleteAccountScreen.kt (hunk 11 above)
Comment (rBergCode) [reply app/src/main/java/ch/hikemate/app/ui/profile/DeleteAccountScreen.kt:55]: It is because this string is used inside of a click function which is not in the composable tree. All others are.
Do you think I should put them all in the same place ?

File: app/src/main/java/ch/hikemate/app/model/authentication/AuthRepository.kt [hunk 17]
```diff
@@ -71,4 +72,29 @@ interface AuthRepository {
...
+   * @param password The password of the user.
+   * @param activity The Android Activity, used for launching the re-authentication dialog.
+   * @param onSuccess Callback to invoke after the user's account has been successfully deleted.
+   * @param onErrorAction Callback to invoke when an error occurs during account deletion. Passes
+   *   the Throwable error.
+   */
+  fun deleteAccount(
+      password: String,
+      activity: Activity,
+      onSuccess: () -> Unit = {},
```
[the remark this answers is not in the document]
Comment (rBergCode) [reply app/src/main/java/ch/hikemate/app/model/authentication/AuthRepository.kt:90]: It was a mistake. Good catch!

File: app/src/main/java/ch/hikemate/app/model/authentication/FirebaseAuthRepository.kt (hunk 3 above)
Comment (rBergCode) [reply app/src/main/java/ch/hikemate/app/model/authentication/FirebaseAuthRepository.kt:165]: I fixed this. Note that there is a sonar warning about `Task<Void>` for which it suggest to change it to `Task<Unit>`. This doesn't work as the tasks types for firebase operations are `Task<Void>`. I suggest that we remove this check in sonar as this is not that bad of a practice.

**Comment by Wylwi** [comment]
@rBergCode I advise you resolve the conflict before I review it again

File: app/src/main/java/ch/hikemate/app/model/authentication/FirebaseAuthRepository.kt [hunk 18]
```diff
@@ -155,24 +157,33 @@ class FirebaseAuthRepository : AuthRepository {
...
+                    .delete()
+
+            val deleteSavedHikesTask =
+                FirebaseFirestore.getInstance()
+                    .collection(SavedHikesRepositoryFirestore.SAVED_HIKES_COLLECTION)
+                    .document(user.uid)
+                    .delete()
+
+            Tasks.whenAll(deleteProfileTask, deleteSavedHikesTask).addOnCompleteListener {
+                deleteDataTask: Task<Void?> ->
```
Comment (Wylwi) [inline app/src/main/java/ch/hikemate/app/model/authentication/FirebaseAuthRepository.kt:173]: Warning: Even if Firebase operations are `Task<Void>`, it's still bad practice for both reasons that Sonar gives us : 
1. Wrong logic
   - Void is not equivalent to Java void but `java.lang.Void.`
2. Unnecessary platform dependency
   -  Void is a platform type available only in the Java Runtime Environment.

Second one is more than an issue for us since we are not running using the JRE but ART, which is a specifically designed runtime for Android (ART is simply short for Android RunTime). Thus, declaring `Task<Void>` or any kind of Void is an issue on non-JRE devices. The void keyword is, here, not related to Firebase tasks. If you look carefully at `.whennAll` documentation, it says that it will always return a `Task<Void>`, which is then handled by the library for the correctness of the implementation. 

Since you don't create an object, and just reuse the library created and implemented ones, there exist a simple fix. Simply remove the type annotation, which is not needed here! Then leave it in the care of the library!

**Review by Wylwi — CHANGES_REQUESTED** [review]
We're on the right track! One last change is needed to fix the Sonar warning, and then this PR can be merged!

**Comment by Wylwi** [comment]
Closes #124

File: app/src/test/java/ch/hikemate/app/authentication/FirebaseAuthRepositoryTest.kt (hunk 7 above)
Comment (rBergCode) [reply app/src/test/java/ch/hikemate/app/authentication/FirebaseAuthRepositoryTest.kt:356]: Then it doesn't check.

File: app/src/main/java/ch/hikemate/app/model/authentication/FirebaseAuthRepository.kt (hunk 18 above)
Comment (rBergCode) [reply app/src/main/java/ch/hikemate/app/model/authentication/FirebaseAuthRepository.kt:173]: Thanks for the explanation and the quick fix!

**Review by Wylwi — APPROVED** [review]
Thanks for the quick fix! You can merge it when the checks are finished running!
